### PR TITLE
RAM Code Refactoring

### PR DESCRIPTION
**Pull request by aquacomputer**
RAM Code Refactoring
Added Virtual Memory

### CODE REVIEW

File: Hardware/RAM/RAMInterop.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// Copyright (C) 2016-2019 Sebastian Grams <https://github.com/sebastian-dev>
+// Copyright (C) 2016-2019 Aqua Computer <https://github.com/aquacomputer, info@aqua-computer.de>
+
+using System;
+using System.IO;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+
+namespace OpenHardwareMonitor.Hardware.RAM {
+  public class Interop {
```
Comment (Phyxion) [inline Hardware/RAM/RAMInterop.cs]: I think we should follow existing structure here and put this in main RAM file under private NativeMethods class.

File: Hardware/RAM/GenericRAM.cs [hunk 2]
```diff
@@ -1,35 +1,42 @@
...
 
 using System.Runtime.InteropServices;
 
 namespace OpenHardwareMonitor.Hardware.RAM {
   internal class GenericRAM : Hardware {
 
-    private Sensor loadSensor;
-    private Sensor usedMemory;
-    private Sensor availableMemory;
+    private Sensor physicalMemoryUsed { get; set; } = null;
```
Comment (Phyxion) [inline Hardware/RAM/GenericRAM.cs]: These can be fields.

File: Hardware/RAM/GenericRAM.cs [hunk 3]
```diff
@@ -1,35 +1,42 @@
-﻿/*
- 
-  This Source Code Form is subject to the terms of the Mozilla Public
-  License, v. 2.0. If a copy of the MPL was not distributed with this
-  file, You can obtain one at http://mozilla.org/MPL/2.0/.
- 
-  Copyright (C) 2012 Michael Möller <mmoeller@openhardwaremonitor.org>
-	
-*/
+﻿// This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (Phyxion) [inline Hardware/RAM/GenericRAM.cs]: I've noticed quite a lot of copyright changes, I would think we should revisit this and add a single reusable copyright statement everywhere which just says copyrighted under MPL and that users should see the LICENSE file for more information (such as https://github.com/dotnet/foundation/blob/master/guidance/copyright.md). What do you think about this? I'm not sure if it's really necessary to add authors to each file as this is already covered by MPL as a Contributor (which is also visible by GitHub).

File: Hardware/RAM/GenericRAM.cs (hunk 2 above)
Comment (aquacomputer) [reply Hardware/RAM/GenericRAM.cs]: with get/set you see in visual studio all cross references and usage of this objects.
The usage of get/set make no difference in performance.

Comment (Phyxion) [reply Hardware/RAM/GenericRAM.cs]: These should be fields, properties are not for that purpose. Unused fields should be highlighted by VS (at least it does here).

Comment (aquacomputer) [reply Hardware/RAM/GenericRAM.cs]: You see when you use get/set all references to all used code parts, the same like methods. 
the auto generated property like this: private Sensor physicalMemoryUsed { get; set; } = null;
is optimized by jit.
But is better readable. Especially when the project gets bigger.

Comment (Phyxion) [reply Hardware/RAM/GenericRAM.cs]: I disagree on the readability part. Properties shouldn't be used for that (and rarely used for that), it's confusing. Properties should be used as if they are public data members, which is not the case for these private implementation details.

Comment (aquacomputer) [reply Hardware/RAM/GenericRAM.cs]: code_preview

Comment (Phyxion) [reply Hardware/RAM/GenericRAM.cs]: I know about CodeLens but that doesn't change anything regarding the readability and purpose of fields versus properties.  You can always right click and select "Find All References" for fields.

File: Hardware/RAM/GenericRAM.cs (hunk 3 above)
Comment (aquacomputer) [reply Hardware/RAM/GenericRAM.cs]: Sounds good.
Whenever something is edited, the old copyright part can be replaced by the new one.

// Mozilla Public License 2.0
// If a copy of the MPL was not distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
// Copyright (c) .NET Foundation and Contributors
// All Rights Reserved

Comment (Phyxion) [reply Hardware/RAM/GenericRAM.cs]: I would change it to LibreHardwareMonitor instead of .NET Foundation but that would be OK then 👍

**Pull request merged**